pkg/registry/core/service/storage/rest.go

```diff
@@ -390,35 +434,32 @@ func (rs *REST) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
 	}
 
 	// TODO: this should probably move to strategy.PrepareForCreate()
-	releaseServiceIP := false
+	var toReleaseClusterIPs map[api.IPFamily]string
+	var toAllocClusterIPs map[api.IPFamily]string
+
+	// failed allocation release
 	defer func() {
-		if releaseServiceIP {
-			if helper.IsServiceIPSet(service) {
-				allocator := rs.getAllocatorByClusterIP(service)
-				allocator.Release(net.ParseIP(service.Spec.ClusterIP))
-			}
+		// release the allocate .. allocate the released
+		released, errRelease := rs.releaseClusterIPs(toReleaseClusterIPs)
+		allocated, errAllocate := rs.allocClusterIPs(oldService, toAllocClusterIPs)
```

Comment:
This is racy - someone else could allocate a released IP in between?  We probably should not release anything until all new allocations are done?